### PR TITLE
Fix flex attention shared memory config fallback

### LINKED ISSUE #133254: Shared memory out of resource when using flex attention
### 🐛 Describe the bug

When I use flex attention on one RTX 4090, I got some error.
A minimal repro:
```python
import torch
from torch.nn.attention.flex_attention import flex_attention
flex_attention = torch.compile(flex_attention, dynamic=False)

B = 1
H = 16
S = 2048
D = 128

query = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
key = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
value = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
gradOut = torch.randn(B, H, S, D, device="cuda", dtype=torch.bfloat16)

flex_out = flex_attention(query, key, value)
flex_out.backward(gradOut, retain_graph=True)
```
It works with D=64 or without compile.
(I see https://github.com/pytorch/pytorch/issues/132075, while this happens during backward)

### Error logs

```
triton.runtime.errors.OutOfResources: out of resource: shared memory, Required: 131074, Hardware limit: 101376. Reducing block sizes or `num_stages` may help.
```

### Versions

torch==2.5.0.dev20240810+cu121

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @drisspg 
```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184310

Add fallback FlexAttention configs for low shared-memory CUDA devices and filter forward/backward configs against the final kernel shared-memory estimate before selecting a non-autotune config. This avoids selecting kernels that compile during benchmarking but fail final compilation on devices such as RTX 4090/3090/A10G.

Fixes #133254

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by drisspg — CHANGES_REQUESTED** [review]
Same

**Comment by jansel** [comment]
@claude review these changes

**Comment by SakshamKapoor2911** [comment]
Applied your approach to current main with a fix for the backward overhead constant (it was 8, I normalized both directions to 4096). Opened #191269 on main -- happy to contribute to your PR instead if you prefer that route.

**Comment by jansel** [comment]
My agent says this stale PR should be closed rather than rebased: current Triton on target sm89 compiles the original D=128 forward/backward kernels within the 101376-byte limit after its built-in stage-1 retry (98304 and 67584 bytes), and no current case was found where smaller blocks are required. The static estimator over-prunes valid configs, while unconditional fallbacks add ordinary-mode compilation overhead. The broader issue can remain open for a concrete current ScoreMod-specific repro.
